common/workunit/workflow.cpp

```diff
@@ -613,12 +707,672 @@ WorkflowMachine::WorkflowMachine(const IContextLogger &_logctx)
+{
+    if(wfid <= workflow->count())
+    {
+        return static_cast<CCloneWorkflowItem&>(workflow->queryWfid(wfid));;
+    }
+    else
+    {
+        unsigned index = wfid - workflow->count() - 1;
+        if(index >= logicalWorkflow.size())
+            throwUnexpected();
+        return static_cast<CCloneWorkflowItem&>(*logicalWorkflow[index].get());
+    }
+}
+void WorkflowMachine::initialiseItemQueue()
+{
+    //Items without any predecessors can be executed by a thread
+    //all runtime workflow items should have predecessors or be contingencies
+    for(int i = 1; i <= workflow->count(); i++)
+    {
+        CCloneWorkflowItem & cur = static_cast<CCloneWorkflowItem&>(workflow->queryWfid(i));
+
+        if(cur.queryNumPredecessors() == 0)
+            itemQueue.push(i);
+    }
+    itemQueueSem.signal(itemQueue.size());
+}
+bool WorkflowMachine::checkIfDone()
+{
+    if((pendingContingencies == 0) && (rootReached))
+    {
+        if( (branchCount == 0) || failureClauseOld)
+        {
+        #ifdef TRACE_WORKFLOW
+                LOG(MCworkflow, "workflow done");
+        #endif
+            done = true;
+            itemQueueSem.signal(numThreads);
+            return true;
+        }
+    }
+    return false;
+}
+void WorkflowMachine::processWfItems()
+{
+    unsigned currentWfid = 0;
+    Owned<WorkflowException> error;
+    while(!done)
+    {
+        itemQueueSem.wait();
+        if(!done)
+        {
+            //enter critical section
+            {
+            CriticalBlock thisBlock(queueCritSec);
+            currentWfid = itemQueue.front();
+            itemQueue.pop();
+            }
+            try
+            {
+                if(!error)
+                {
+                    executeItemParallel(currentWfid);
+                    currentWfid = 0;
+                }
+            }
+            catch(WorkflowException * e)
+            {
+                error.setown(e);
+                break;
+            }
+        }
+    }
+    if(error)
+        throw error.getLink();
+}
+void WorkflowMachine::performParallel(IGlobalCodeContext *_ctx, IEclProcess *_process)
+{
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "starting perform parallel");
+    #endif
+    ctx = _ctx;
+    process = _process;
+
+    //relink workflow
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Starting to mark Items with their successors");
+    #endif
+    addSuccessors();
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Finished marking Items with their successors");
+    #endif
+
+    //add initial values
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Adding initial workflow items");
+    #endif
+    initialiseItemQueue();
+
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Initialising threads");
+    #endif
+
+    //initialise thread count
+    getThreadNumFlag();
+
+    std::vector<std::thread *> threads(numThreads);
+    for(int i=0; i < numThreads; i++)
+    {
```

Comment:
optional, but possibly clearer when there is only a single statement inside a for to not have the {}